File: bench/agent-efficiency/fixtures/python/ledger/store.py
"""Tab-separated persistence for a ledger."""
from __future__ import annotations

from datetime import datetime
from typing import IO

from .ledger import Entry, Ledger, apply_entry, new_ledger


def save(stream: IO[str], ledger: Ledger) -> None:
    """Write the ledger as tab-separated lines: id, account, amount, currency,
    category, posted (ISO 8601), memo."""
    for entry in ledger.entries:
        assert entry.posted is not None
        fields = [
            entry.id, entry.account, str(entry.amount), entry.currency, entry.category,
            entry.posted.isoformat(), entry.memo,
        ]
        stream.write("\t".join(fields) + "\n")


def load(stream: IO[str]) -> Ledger:
    """Read lines written by save and apply them to a new ledger."""
    ledger = new_ledger()
    for line_no, line in enumerate(stream, start=1):
        fields = line.rstrip("\n").split("\t")
        if len(fields) != 7:
            raise ValueError(f"store: line {line_no} has {len(fields)} fields, want 7")
        entry = Entry(
            id=fields[0], account=fields[1], amount=int(fields[2]), currency=fields[3],
            category=fields[4], posted=datetime.fromisoformat(fields[5]), memo=fields[6],
        )
        apply_entry(ledger, entry)
    return ledger

File: bench/agent-efficiency/fixtures/go/store.go
package ledger

import (
	"bufio"
	"fmt"
	"io"
	"strconv"
	"strings"
	"time"
)

// Save writes the ledger as tab-separated lines: id, account, amount,
// currency, category, posted (RFC 3339), memo.
func Save(w io.Writer, l *Ledger) error {
	bw := bufio.NewWriter(w)
	for _, e := range l.Entries() {
		line := strings.Join([]string{
			e.ID, e.Account, strconv.FormatInt(e.Amount, 10), e.Currency, e.Category,
			e.Posted.UTC().Format(time.RFC3339), e.Memo,
		}, "\t")
		if _, err := bw.WriteString(line + "\n"); err != nil {
			return err
		}
	}
	return bw.Flush()
}

// Load reads lines written by Save and applies them to a new ledger.
func Load(r io.Reader) (*Ledger, error) {
	l := NewLedger()
	scanner := bufio.NewScanner(r)
	lineNo := 0
	for scanner.Scan() {
		lineNo++
		fields := strings.Split(scanner.Text(), "\t")
		if len(fields) != 7 {
			return nil, fmt.Errorf("store: line %d has %d fields, want 7", lineNo, len(fields))
		}
		amount, err := strconv.ParseInt(fields[2], 10, 64)
		if err != nil {
			return nil, fmt.Errorf("store: line %d amount: %w", lineNo, err)
		}
		posted, err := time.Parse(time.RFC3339, fields[5])
		if err != nil {
			return nil, fmt.Errorf("store: line %d posted: %w", lineNo, err)
		}
		e := Entry{ID: fields[0], Account: fields[1], Amount: amount, Currency: fields[3], Category: fields[4], Posted: posted, Memo: fields[6]}
		if err := ApplyEntry(l, e); err != nil {
			return nil, fmt.Errorf("store: line %d: %w", lineNo, err)
		}
	}
	return l, scanner.Err()
}
